cmd/hydrate/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+use humility::hubris::HubrisFlashMap;
+use humility_arch_arm::ARMRegister;
+use humility_cli::{ExecutionContext, Subcommand};
+use humility_cmd::{Archive, Command, CommandKind};
+use humility_log::msg;
+use std::{collections::BTreeMap, io::Read, path::PathBuf};
+
+#[derive(Parser, Debug)]
+#[clap(
+    name = "hydrate", about = env!("CARGO_PKG_DESCRIPTION"),
+    group = ArgGroup::new("target").multiple(false)
+)]
+struct Args {
+    /// Path to the raw memory dump
+    file: PathBuf,
+}
+
+struct DryCore {
+    flash: HubrisFlashMap,
+    mem: BTreeMap<u32, Vec<u8>>,
+}
+
+// Helper macro to stub out functions
+macro_rules! unsupported{
+    ($fn_name:ident($($arg_name:ident: $arg_type:ty),*)) => {
+        unsupported!($fn_name($($arg_name: $arg_type),*) -> Result<()>);
+    };
+    ($fn_name:ident($($arg_name:ident: $arg_type:ty),*) -> $out:ty) => {
+        fn $fn_name(&mut self, $($arg_name: $arg_type),*) -> $out {
+            bail!(concat!(
+                "DryCore does not support ",
+                stringify!($fn_name)))
+        }
+    };
+}
+
+impl humility::core::Core for DryCore {
+    unsupported!(run());
+    unsupported!(halt());
+    unsupported!(step());
+    unsupported!(load(_path: &std::path::Path));
+    unsupported!(reset());
+    unsupported!(write_8(_addr: u32, _data: &[u8]));
+    unsupported!(op_done());
+    unsupported!(op_start());
+    unsupported!(read_reg(_reg: ARMRegister) -> Result<u32>);
+    unsupported!(init_swv());
+    unsupported!(read_swv() -> Result<Vec<u8>>);
+    unsupported!(write_reg(_reg: ARMRegister, _value: u32));
+    unsupported!(write_word_32(_addr: u32, _data: u32));
+    unsupported!(reset_and_halt(_dur: std::time::Duration));
+    unsupported!(wait_for_halt(_dur: std::time::Duration));
+
+    fn info(&self) -> (String, Option<String>) {
+        ("DryCore".to_owned(), None)
+    }
+    fn is_archive(&self) -> bool {
+        false
+    }
+    fn is_dump(&self) -> bool {
+        true // I guess?
+    }
+    fn is_net(&self) -> bool {
+        false
+    }
+    fn vid_pid(&self) -> Option<(u16, u16)> {
+        None
+    }
+
+    fn read_8(&mut self, addr: u32, data: &mut [u8]) -> Result<()> {
+        if self.flash.read(addr, data).is_some() {
+            return Ok(());
+        }
+
+        let Some((base, mem)) = self.mem.range(0..=addr).next_back() else {
+            bail!("addr {addr:#08x} is below memory range");
+        };
+        let offset = (addr - base) as usize;
+        let end = offset + data.len();
+        if end > mem.len() {
+            bail!("region is not large enough; {end:#x} > {:#x}", mem.len());
+        }
+        data.copy_from_slice(&mem[offset..end]);
+        Ok(())
+    }
+}
+
+fn run(context: &mut ExecutionContext) -> Result<()> {
+    let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
+    let subargs = Args::try_parse_from(subargs)?;
+    let f = std::fs::File::open(&subargs.file)?;
+    let mut z = zip::ZipArchive::new(f)?;
```

Comment:
How well documented/stable is this format and how mad are we going to be if we need to change it in the future?

Reply:
I don't expect it to change much, because there's only so many ways to represent a bunch of bytes.  Still, I added `meta.json` containing a `version` value.

Reply:
it occurs to me that if we want to make it a _little_ easier to evolve the format in the future, we could maybe stuff the paths to the various file entries we read out of the zip file in `meta.json`, so the format is a little bit more self-describing? i.e. instead of `z.by_name("TASK_INDEX")`, you would do `z.by_name(meta.task_index_path)` and so on. 

that way, if the paths change, we don't actually need to treat that as a totally new version that has to be handled separately. on the other hand, i'm not sure if changing these paths is actually a thing that will, like, ever reasonably happen, so maybe this isn't worth the effort, even if it's not that much effort.

Reply:
I decided to move the metadata from scattered files in the ZIP archive into the JSON file (renamed to `dump.json`), which lets us deserialize with `serde` (and makes the questions about trailing newlines moot)

Reply:
lovely!